src/script/common/c_content.cpp

```diff
@@ -811,8 +811,30 @@ ItemStack read_item(lua_State* L, int index,Server* srv)
 		std::string name = getstringfield_default(L, index, "name", "");
 		int count = getintfield_default(L, index, "count", 1);
 		int wear = getintfield_default(L, index, "wear", 0);
-		std::string metadata = getstringfield_default(L, index, "metadata", "");
-		return ItemStack(name, count, wear, metadata, idef);
+
+		ItemStack istack=ItemStack(name, count, wear, idef);
+
+		lua_getfield(L, index, "metadata");
+		int fieldstable = lua_gettop(L);
+
+		//if given a table, assume new metadata format, else mimic legacy behavior
+		if(lua_istable(L, fieldstable)){
+			lua_pushnil(L);
+			while(lua_next(L, fieldstable) != 0){
+				// key at index -2 and value at index -1
+				std::string name = lua_tostring(L, -2);
+				size_t cl;
+				const char *cs = lua_tolstring(L, -1, &cl);
+				std::string value(cs, cl);
+				istack.metadata.setString(name, value);
+				lua_pop(L, 1); // removes value, keeps key for next iteration
+			}
+		}else{
+			std::string value = getstringfield_default(L, index, "metadata", "");
+			istack.metadata.setString("", value);
+		}
```

Comment:
Add spaces (`if(` -> `if (`, `}else{` -> `} else {`, `while(` -> `while (`).
Also, use `lua_tolstring` for both key/name and value.